File: frontend/src/api/imm11.ts
// Copyright (c) 2026, AssetCore Team
// API client cho Module IMM-11 — Calibration

import { frappeGet, frappePost } from './helpers'

export interface CalibrationSchedule {
  name: string
  asset: string
  asset_name?: string
  device_model: string
  calibration_type: 'External' | 'In-House'
  interval_days: number
  last_calibration_date: string | null
  next_due_date: string | null
  preferred_lab: string | null
  is_active: 0 | 1
}

export interface CalibrationMeasurement {
  idx?: number
  parameter_name: string
  unit: string
  nominal_value: number
  tolerance_positive: number
  tolerance_negative: number
  measured_value: number | null
  out_of_tolerance?: 0 | 1
  pass_fail?: 'Pass' | 'Fail' | null
}

/**
 * Raw measurement fields KTV nhập — tập DUY NHẤT được gửi lên BE khi lưu phiếu.
 * `pass_fail` / `out_of_tolerance` do SERVER tính (SSoT = controller cha
 * `_compute_measurement_results`, imm_asset_calibration.py) → FE KHÔNG BAO GIỜ gửi
 * hai field này (không tin payload client), chỉ render lại sau khi reload. Dùng
 * `Pick` để lệ thuộc DUY NHẤT nguồn field CalibrationMeasurement.
 */
export type CalibrationMeasurementInput = Pick<
  CalibrationMeasurement,
  'parameter_name' | 'unit' | 'nominal_value'
  | 'tolerance_positive' | 'tolerance_negative' | 'measured_value'
>

/**
 * Patch cho update_calibration — CHỈ field scalar editable (mirror BE
 * `_UPDATE_ALLOWED`) + `measurements` child-diff (raw-only). CÓ key `measurements`
 * ⇒ BE replace-set theo parameter_name/idx + tính lại pass_fail server-side;
 * VẮNG key ⇒ backward-compat scalar-only (hành vi cũ, 0 regression caller cũ).
 * Transport: `updateCalibration` JSON.stringify `measurements` theo convention
 * imm08/imm09 (BE `parse_json`).
 */
export interface CalibrationUpdatePatch {
  status?: AssetCalibration['status']
  actual_date?: string | null
  lab_supplier?: string | null
  lab_accreditation_number?: string | null
  lab_contract_ref?: string | null
  sent_date?: string | null
  certificate_number?: string | null
  certificate_date?: string | null
  reference_standard_serial?: string | null
  traceability_reference?: string | null
  technician_notes?: string | null
  amendment_reason?: string | null
  measurements?: CalibrationMeasurementInput[]
}

export interface AssetCalibration {
  name: string
  asset: string
  asset_name?: string
  device_model: string
  calibration_schedule: string | null
  calibration_type: 'External' | 'In-House'
  status: 'Scheduled' | 'Sent to Lab' | 'In Progress' | 'Certificate Received' | 'Passed' | 'Failed' | 'Conditionally Passed' | 'Cancelled'
  scheduled_date: string
  actual_date: string | null
  technician: string
  technician_name?: string
  assigned_by: string | null
  lab_supplier: string | null
  lab_accreditation_number: string | null
  lab_contract_ref: string | null
  sent_date: string | null
  sent_by: string | null
  certificate_file: string | null
  certificate_date: string | null
  certificate_number: string | null
  next_calibration_date: string | null
  overall_result: 'Passed' | 'Failed' | 'Conditionally Passed' | null
  reference_standard_serial: string | null
  traceability_reference: string | null
  measurements: CalibrationMeasurement[]
  pm_work_order: string | null
  capa_record: string | null
  is_recalibration: 0 | 1
  calibration_sticker_attached: 0 | 1
  sticker_photo: string | null
  technician_notes: string | null
  amendment_reason: string | null
  docstatus?: 0 | 1 | 2
  // SoT server-driven CTA: tập trạng-thái-kế hợp lệ per status từ BE
  // (_CAL_VALID_TRANSITIONS, imm11.py:1033). FE gate nút workflow theo list này —
  // KHÔNG hardcode status→button client-side (anti-pattern dead-gate/flow-drift).
  allowed_transitions?: string[]
  // Cờ hạn hiệu chuẩn derive SERVER-SIDE (CR-02 · server-flag SSoT). list_calibrations
  // + get_calibration emit is_overdue/is_due_soon (int 0/1) qua CHUNG helper BE
  // is_calibration_overdue / is_calibration_due_soon (services/imm11.py). Consumer CHỈ
  // render cờ (calFlagBadge), KHÔNG so next_calibration_date với client-clock. Overdue
  // ưu tiên due_soon; None/chưa-có-hạn → cả hai 0. INV parity list==detail (INV-SLA-5).
  is_overdue?: 0 | 1
  is_due_soon?: 0 | 1
}

export interface CalibrationKpis {
  kpis: {
    total_this_month: number
    completed: number
    failed: number
    pass_rate_pct: number
    overdue_assets: number
    due_soon_assets: number
  }
}

const BASE = '/api/method/assetcore.api.imm11'

export async function listCalibrationSchedules(filters = {}, page = 1, pageSize = 20) {
  // Tier 2 service trả {data, pagination}
  return frappeGet<{ data: CalibrationSchedule[]; pagination: Record<string, number> }>(
    `${BASE}.list_calibration_schedules`, { filters: JSON.stringify(filters), page, page_size: pageSize },
  )
}

export async function getCalibrationSchedule(name: string) {
  return frappeGet<CalibrationSchedule>(`${BASE}.get_calibration_schedule`, { name })
}

export async function createCalibrationSchedule(payload: Partial<CalibrationSchedule>) {
  return frappePost<{ name: string; next_due_date: string }>(`${BASE}.create_calibration_schedule`, payload as Record<string, unknown>)
}

export async function updateCalibrationSchedule(name: string, data: Partial<CalibrationSchedule>) {
  return frappePost<{ name: string }>(`${BASE}.update_calibration_schedule`, { name, ...data } as Record<string, unknown>)
}

export async function deleteCalibrationSchedule(name: string) {
  return frappePost<{ name: string; deleted: boolean }>(`${BASE}.delete_calibration_schedule`, { name })
}

export async function listCalibrations(filters = {}, page = 1, pageSize = 20) {
  // Tier 2 service trả {data, pagination}
  return frappeGet<{ data: AssetCalibration[]; pagination: Record<string, number> }>(
    `${BASE}.list_calibrations`, { filters: JSON.stringify(filters), page, page_size: pageSize },
  )
}

export async function getCalibration(name: string) {
  return frappeGet<AssetCalibration>(`${BASE}.get_calibration`, { name })
}

export async function createCalibration(payload: {
  asset: string
  calibration_type: string
  scheduled_date: string
  technician: string
  calibration_schedule?: string
  lab_supplier?: string
  is_recalibration?: number
}) {
  return frappePost<{ name: string; status: string }>(`${BASE}.create_calibration`, payload as Record<string, unknown>)
}

export async function updateCalibration(name: string, data: CalibrationUpdatePatch) {
  // Serialize child-diff theo convention imm08/imm09: nested-array param = JSON string,
  // BE `parse_json`. CHỈ stringify khi CÓ key `measurements` ⇒ giữ backward-compat
  // (vắng key = scalar-only, không đụng bảng con) — caller scalar (vd doStartCal chỉ gửi
  // {status}) KHÔNG kèm `measurements`, đi đúng nhánh cũ.
  const body: Record<string, unknown> = { name, ...data }
  if (data.measurements !== undefined) {
    body.measurements = JSON.stringify(data.measurements)
  }
  return frappePost<{ name: string; status: string }>(`${BASE}.update_calibration`, body)
}

export async function submitCalibration(name: string) {
  return frappePost<{ name: string; status: string; overall_result: string; next_calibration_date: string }>(
    `${BASE}.submit_calibration`, { name },
  )
}

export async function getCalibrationKpis(year?: number, month?: number) {
  return frappeGet<CalibrationKpis>(`${BASE}.get_calibration_kpis`, { year, month })
}

export async function getAssetCalibrationHistory(asset: string, limit = 10) {
  return frappeGet<{ asset: string; history: AssetCalibration[] }>(
    `${BASE}.get_asset_calibration_history`, { asset, limit },
  )
}

export async function sendToLab(name: string, payload: {
  sent_date?: string; lab_supplier?: string; lab_contract_ref?: string
} = {}) {
  return frappePost<{ name: string; status: string; sent_date: string }>(
    `${BASE}.send_to_lab`, { name, ...payload } as Record<string, unknown>,
  )
}

export async function receiveCertificate(name: string, payload: {
  certificate_file: string
  certificate_number: string
  certificate_date: string
  traceability_reference?: string
  reference_standard_serial?: string
}) {
  return frappePost<{ name: string; status: string; certificate_number: string }>(
    `${BASE}.receive_certificate`, { name, ...payload } as Record<string, unknown>,
  )
}

export async function cancelCalibration(name: string, reason: string) {
  return frappePost<{ name: string; status: string }>(
    `${BASE}.cancel_calibration`, { name, reason },
  )
}

export interface DueCalibrationItem {
  name: string
  asset_name: string
  device_model: string
  location: string | null
  next_calibration_date: string | null
  calibration_status: string | null
  days_left: number | null
}

export async function getDueCalibrations(days = 30, limit = 50) {
  return frappeGet<{ items: DueCalibrationItem[]; threshold_days: number }>(
    `${BASE}.get_due_calibrations`, { days, limit },
  )
}

File: assetcore/api/imm11.py
# Copyright (c) 2026, AssetCore Team
# REST API cho Module IMM-11 — Calibration.
#
# Tier 1 — Presentation only.
# Parse HTTP input → gọi service → format _ok / _err envelope.
# KHÔNG gọi frappe.db.* hay frappe.get_doc trực tiếp.

from __future__ import annotations

import datetime

import frappe

from assetcore.services import imm11 as svc
from assetcore.services.shared import ServiceError
from assetcore.services.shared import rbac
from assetcore.services.shared.scope import apply_vendor_scope, assert_vendor_can_access
from assetcore.utils.api_handler import handle, parse_json
from assetcore.utils.response import _err


# ─── 1. Calibration Schedules ────────────────────────────────────────────────

@frappe.whitelist()
def list_calibration_schedules(filters: str = "{}", page: int = 1, page_size: int = 20) -> dict:
    try:
        f = parse_json(filters, default={}, field_name="filters")
    except ServiceError as e:
        return _err(e.message, e.code, http_status=e.http_status)
    f = apply_vendor_scope(f, "Calibration Schedule")
    return handle(svc.list_schedules, f, page=int(page), page_size=int(page_size))


@frappe.whitelist()
def get_calibration_schedule(name: str) -> dict:
    try:
        assert_vendor_can_access("Calibration Schedule", name)
    except ServiceError as e:
        return _err(e.message, e.code, http_status=e.http_status)
    return handle(svc.get_schedule, name)


@frappe.whitelist()
def create_calibration_schedule(asset: str, calibration_type: str, interval_days: int,
                                 preferred_lab: str = None, next_due_date: str = None) -> dict:
    # AUTH-02 — server-side gate; FE button hiding is not a security control.
    rbac.require("calibration.create")
    return handle(
        svc.create_schedule,
        asset=asset, calibration_type=calibration_type,
        interval_days=int(interval_days),
        preferred_lab=preferred_lab, next_due_date=next_due_date,
    )


@frappe.whitelist()
def update_calibration_schedule(name: str, **kwargs) -> dict:
    rbac.require("calibration.write")
    return handle(svc.update_schedule, name, kwargs)


@frappe.whitelist()
def delete_calibration_schedule(name: str) -> dict:
    rbac.require("calibration.delete")
    return handle(svc.delete_schedule, name)


# ─── 2. Calibration Work Orders ───────────────────────────────────────────────

@frappe.whitelist()
def list_calibrations(filters: str = "{}", mine: int = 0,
                      page: int = 1, page_size: int = 20) -> dict:
    # C-LISTREAD-MINE-CAL (quartet "phiếu-của-tôi" ĐÓNG NỐT sau PM/CM/Incident): tab "Phiếu hiệu
    #   chuẩn của tôi" (MVP-5d) truyền mine=1 → scope technician == session.user (calibration
    #   assignee — KHÔNG assigned_to; mirror IncidentMine dùng reported_by, mỗi domain field RIÊNG).
    #   Inject SAU apply_vendor_scope (vendor-scope vẫn áp trước). mine=0/absent ⇒ filters
    #   byte-identical baseline (web-FE list_calibrations KHÔNG đổi). count==rows giữ: count_with_or
    #   + get_all dùng CÙNG filters dict (đã có technician). Mirror imm08.py:28 / imm09.py:22.
    try:
        f = parse_json(filters, default={}, field_name="filters")
    except ServiceError as e:
        return _err(e.message, e.code, http_status=e.http_status)
    f = apply_vendor_scope(f, "Calibration Record")
    if int(mine or 0):
        f["technician"] = frappe.session.user
    return handle(svc.list_calibrations, f, page=int(page), page_size=int(page_size))


@frappe.whitelist()
def get_calibration(name: str) -> dict:
    try:
        assert_vendor_can_access("Calibration Record", name)
    except ServiceError as e:
        return _err(e.message, e.code, http_status=e.http_status)
    return handle(svc.get_calibration, name)


@frappe.whitelist(methods=["POST"])
def create_calibration(asset: str, calibration_type: str, scheduled_date: str,
                        technician: str, calibration_schedule: str = None,
                        lab_supplier: str = None, is_recalibration: int = 0,
                        reference_standard_serial: str = None,
                        traceability_reference: str = None) -> dict:
    rbac.require("calibration.create")
    return handle(
        svc.create_calibration,
        asset=asset, calibration_type=calibration_type,
        scheduled_date=scheduled_date, technician=technician,
        calibration_schedule=calibration_schedule,
        lab_supplier=lab_supplier,
        is_recalibration=int(is_recalibration),
        reference_standard_serial=reference_standard_serial,
        traceability_reference=traceability_reference,
    )


@frappe.whitelist()
def update_calibration(name: str, **kwargs) -> dict:
    rbac.require("calibration.write")
    # CR-24-WEB: `measurements` (child-diff nhập-đo web) đến dạng list (JSON body) HOẶC
    # JSON-string (form-encoded — mobile_be form_dict oneOf json+form). parse_json idempotent
    # cho list ⇒ chuẩn hoá về list-of-dict trước khi xuống service. parse_json malformed →
    # ServiceError(INVALID_PARAMS) envelope (KHÔNG traceback 500).
    if "measurements" in kwargs:
        kwargs["measurements"] = parse_json(
            kwargs["measurements"], default=[], field_name="measurements")
    return handle(svc.update_calibration, name, kwargs)


@frappe.whitelist(methods=["POST"])
def submit_calibration(name: str, client_request_id: str = "") -> dict:
    # CR-24-CAL-SUBMIT (op#6): client_request_id optional (mobile write-outbox idempotency;
    # body THẮNG header). Default str="" (KHÔNG None → tránh 417). rbac.require GIỮ.
    rbac.require("calibration.submit")
    return handle(svc.submit_calibration, name,
                  client_request_id=str(client_request_id or ""))


@frappe.whitelist(methods=["POST"])
def add_measurement(name: str, parameter_name: str, unit: str, nominal_value: float,
                     tolerance_positive: float, tolerance_negative: float,
                     measured_value: float = None, client_request_id: str = "") -> dict:
    rbac.require("calibration.write")
    return handle(
        svc.add_measurement, name,
        parameter_name=parameter_name, unit=unit,
        nominal_value=float(nominal_value),
        tolerance_positive=float(tolerance_positive),
        tolerance_negative=float(tolerance_negative),
        measured_value=float(measured_value) if measured_value is not None else None,
        client_request_id=str(client_request_id or ""),
    )


# ─── 3. KPIs ─────────────────────────────────────────────────────────────────

@frappe.whitelist()
def get_calibration_kpis(year: int = None, month: int = None) -> dict:
    now = datetime.date.today()
    return handle(
        svc.get_kpis,
        int(year) if year else now.year,
        int(month) if month else now.month,
    )


@frappe.whitelist()
def get_calibration_dashboard() -> dict:
    return handle(svc.get_dashboard)


@frappe.whitelist()
def get_asset_calibration_history(asset: str, limit: int = 10) -> dict:
    return handle(svc.get_asset_history, asset, int(limit))


# ─── 4. Workflow actions ─────────────────────────────────────────────────────

@frappe.whitelist(methods=["POST"])
def send_to_lab(name: str, sent_date: str = None, lab_supplier: str = None,
                lab_contract_ref: str = None) -> dict:
    rbac.require("cal.send_lab")
    return handle(
        svc.send_to_lab, name,
        sent_date=sent_date, lab_supplier=lab_supplier,
        lab_contract_ref=lab_contract_ref,
    )


@frappe.whitelist(methods=["POST"])
def receive_certificate(name: str, certificate_file: str,
                        certificate_number: str, certificate_date: str,
                        traceability_reference: str = None,
                        reference_standard_serial: str = None) -> dict:
    rbac.require("calibration.write")
    return handle(
        svc.receive_certificate, name,
        certificate_file=certificate_file,
        certificate_number=certificate_number,
        certificate_date=certificate_date,
        traceability_reference=traceability_reference,
        reference_standard_serial=reference_standard_serial,
    )


@frappe.whitelist(methods=["POST"])
def cancel_calibration(name: str, reason: str) -> dict:
    rbac.require("calibration.cancel")
    return handle(svc.cancel_calibration, name, reason)


@frappe.whitelist()
def get_due_calibrations(days: int = 30, limit: int = 50) -> dict:
    return handle(svc.get_due_calibrations, int(days), int(limit))
